src/results.cpp

```diff
@@ -690,9 +693,21 @@ void Results::prepare_async()
     _impl::RealmCoordinator::register_notifier(m_notifier);
 }
 
-NotificationToken Results::add_notification_callback(CollectionChangeCallback cb) &
+#if REALM_ENABLE_SYNC
+NotificationToken Results::add_notification_callback(CollectionChangeCallback cb, util::Optional<std::string> subscription_name) &
+#else 
+NotificationToken Results::add_notification_callback(CollectionChangeCallback cb, util::Optional<std::string>) &
+#endif
 {
     prepare_async();
+#if REALM_ENABLE_SYNC
+    if (m_realm->is_partial() && !m_have_subscribed) {
+        std::string key = (subscription_name) ? subscription_name.value() : partial_sync::get_default_name(get_query());
```

Comment:
`get_query()` does a non-trivial amount of work, including copying the `Query`. We should use a local variable for the result to avoid doing this work multiple times.